src/storage/hummock_sdk/src/table_watermark.rs

```diff
@@ -208,46 +186,74 @@ impl TableWatermarksIndex {
-                    assert!(*prev_epoch < epoch);
-                    match self.watermark_direction {
-                        WatermarkDirection::Ascending => {
-                            assert!(vnode_watermark.watermark >= prev_watermark);
-                        }
-                        WatermarkDirection::Descending => {
-                            assert!(vnode_watermark.watermark <= prev_watermark);
-                        }
-                    };
-                };
-                assert!(self
-                    .index
-                    .entry(vnode)
-                    .or_default()
-                    .insert(epoch, vnode_watermark.watermark.clone())
-                    .is_none());
+        let mut vnode_is_set = BitmapBuilder::zeroed(VirtualNode::COUNT);
+        #[cfg(debug_assertions)]
+        {
+            for vnode_watermark in vnode_watermark_list.as_ref() {
+                for vnode in vnode_watermark.vnode_bitmap.iter_ones() {
+                    assert!(!vnode_is_set.is_set(vnode));
+                    vnode_is_set.set(vnode, true);
+                    let vnode = VirtualNode::from_index(vnode);
+                    if let Some(prev_watermark) = self.latest_watermark(vnode) {
+                        match self.watermark_direction {
+                            WatermarkDirection::Ascending => {
+                                assert!(vnode_watermark.watermark >= prev_watermark);
+                            }
+                            WatermarkDirection::Descending => {
+                                assert!(vnode_watermark.watermark <= prev_watermark);
+                            }
+                        };
+                    }
+                }
             }
         }
+        self.watermarks.push_back((epoch, vnode_watermark_list));
     }
 
-    pub fn apply_committed_watermarks(&mut self, committed_index: &TableWatermarksIndex) {
-        self.committed_epoch = committed_index.committed_epoch;
-        for (vnode, committed_epoch_watermark) in &committed_index.index {
-            let epoch_watermark = self.index.entry(*vnode).or_default();
-            // keep only watermark higher than committed epoch
-            *epoch_watermark = epoch_watermark.split_off(&committed_index.committed_epoch);
-            for (epoch, watermark) in committed_epoch_watermark {
-                epoch_watermark.insert(*epoch, watermark.clone());
+    pub fn apply_committed_watermarks(
+        &mut self,
+        committed_watermark: &TableWatermarks,
+        committed_epoch: HummockEpoch,
+    ) {
+        assert_eq!(self.watermark_direction, committed_watermark.direction);
+        assert!(self.committed_epoch <= committed_epoch);
+        if self.committed_epoch == committed_epoch {
+            return;
+        }
+        self.committed_epoch = committed_epoch;
+        // keep only watermark higher than committed epoch
+        while let Some((old_epoch, _)) = self.watermarks.front() {
```

Comment:
Given that `watermarks` are sorted by epoch, will it be more efficient to use `binary_search` + `split_off` or `truncate` (need to swtich to `push_front`) here?